### ISSUE TITLE
Manage unit test output directory

### ISSUE THREAD

**Issue by nomis** [mesonbuild/meson#4299]
If unit tests need to write files to perform their tests, they need a separate directory within the build directory to do this. There's currently no defined output directory for unit tests.

Meson should create a known output directory that tests can use (but it would be up to the tests themselves to use unique filenames within this directory). Otherwise tests could interfere with files that Meson wants to create in the build directory (now or in the future).

Meson should also delete and recreate this output directory before running the tests. Otherwise test runs can be affected by files from previous test runs. (This is particularly annoying when you rename a test output file in some places but not others, and it keeps working until you delete your build directory.)

**Comment by jpakkane** [comment]
Every operating system has a temp directory (such as `/tmp`) exactly for this kind of use. Is there some use case it is not catering to?

**Comment by nomis** [comment]
Yes, using it requires me to create a random directory. If I interrupt the unit tests then those files won't be deleted. If I want to look at the files it creates to determine why a test is not working then I can't have the test try and delete them. It's much more convenient to use the build directory.
